pkg/volume/testing/testing.go

```diff
@@ -81,47 +82,48 @@ type fakeVolumeHost struct {
 	subpather       subpath.Interface
 	csiDriverLister storagelisters.CSIDriverLister
 	informerFactory informers.SharedInformerFactory
+	kubeletErr      error
+	mux             sync.Mutex
 }
 
 var _ VolumeHost = &fakeVolumeHost{}
 var _ AttachDetachVolumeHost = &fakeVolumeHost{}
 
 func NewFakeVolumeHost(rootDir string, kubeClient clientset.Interface, plugins []VolumePlugin) *fakeVolumeHost {
-	return newFakeVolumeHost(rootDir, kubeClient, plugins, nil, nil)
+	return newFakeVolumeHost(rootDir, kubeClient, plugins, nil, nil, "", nil)
 }
 
 func NewFakeVolumeHostWithCloudProvider(rootDir string, kubeClient clientset.Interface, plugins []VolumePlugin, cloud cloudprovider.Interface) *fakeVolumeHost {
-	return newFakeVolumeHost(rootDir, kubeClient, plugins, cloud, nil)
+	return newFakeVolumeHost(rootDir, kubeClient, plugins, cloud, nil, "", nil)
 }
 
 func NewFakeVolumeHostWithNodeLabels(rootDir string, kubeClient clientset.Interface, plugins []VolumePlugin, labels map[string]string) *fakeVolumeHost {
-	volHost := newFakeVolumeHost(rootDir, kubeClient, plugins, nil, nil)
+	volHost := newFakeVolumeHost(rootDir, kubeClient, plugins, nil, nil, "", nil)
 	volHost.nodeLabels = labels
 	return volHost
 }
 
 func NewFakeVolumeHostWithCSINodeName(rootDir string, kubeClient clientset.Interface, plugins []VolumePlugin, nodeName string, driverLister storagelisters.CSIDriverLister) *fakeVolumeHost {
-	volHost := newFakeVolumeHost(rootDir, kubeClient, plugins, nil, nil)
-	volHost.nodeName = nodeName
-	if driverLister != nil {
-		volHost.csiDriverLister = driverLister
-	}
+	volHost := newFakeVolumeHost(rootDir, kubeClient, plugins, nil, nil, nodeName, driverLister)
 	return volHost
 }
 
-func newFakeVolumeHost(rootDir string, kubeClient clientset.Interface, plugins []VolumePlugin, cloud cloudprovider.Interface, pathToTypeMap map[string]hostutil.FileType) *fakeVolumeHost {
-	host := &fakeVolumeHost{rootDir: rootDir, kubeClient: kubeClient, cloud: cloud}
+func newFakeVolumeHost(rootDir string, kubeClient clientset.Interface, plugins []VolumePlugin, cloud cloudprovider.Interface, pathToTypeMap map[string]hostutil.FileType, nodeName string, driverLister storagelisters.CSIDriverLister) *fakeVolumeHost {
+	host := &fakeVolumeHost{rootDir: rootDir, kubeClient: kubeClient, cloud: cloud, nodeName: nodeName, csiDriverLister: driverLister}
 	host.mounter = mount.NewFakeMounter(nil)
 	host.hostUtil = hostutil.NewFakeHostUtil(pathToTypeMap)
 	host.exec = &testingexec.FakeExec{DisableScripts: true}
+	host.pluginMgr = &VolumePluginMgr{}
 	host.pluginMgr.InitPlugins(plugins, nil /* prober */, host)
 	host.subpather = &subpath.FakeSubpath{}
 	host.informerFactory = informers.NewSharedInformerFactory(kubeClient, time.Minute)
+	// Wait until the InitPlugins setup is finished before returning from this setup func
+	host.WaitForKubeletErrNil()
```

Comment:
this is ignoring an error return... should it?

Reply:
technically... no. But this func is already ignoring errors from `host.pluginMgr.InitPlugins` and this func has way too many call sites and several layers before a `t *testing.T` is available to throw an error (or to pipe the `t` down to). I explicitly didn't do that to keep the PR size down. My plan was to actually come in and refactor this whole thing after codefreeze (so much duplicated code and weirdness). I can create an issue for it or if you care deeply I can fix it now :)

Reply:
it's unclear to me why we're calling this if we don't care whether it succeeds or not

Reply:
it fixes a data race (it does actually block the goroutine until success in practice). In the failure case it just delays the inevitable failure down the line when something else blows up.

I'll fix it. Will do in a seperate commit since it's going to be large

Reply:
if failure here should fail whatever is calling it, and it's only called in tests, a panic is an ok interim step until callers can be updated.